File: built-in plugins dsp/mb-wt-pluck.cpp
/**
 * MB WT Pluck
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Sharp wavetable pluck
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_PLUCK_H
#define MB_WT_PLUCK_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtPluck : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-pluck";
    static constexpr const char* PLUGIN_NAME    = "MB WT Pluck";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0.5f;  // range [0, 1]
    float attack = 0.8f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbWtPluck() = default;
    ~MbWtPluck() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.attack = std::clamp(params.attack, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB WT Pluck
        return input;
    }
};

#endif // MB_WT_PLUCK_H

File: built-in plugins dsp/mb-wt-pluck.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbWtPluckPlugin: PluginDefinition = { id: 'mb-wt-pluck', slug: 'mb-wt-pluck', name: 'MB WT Pluck', category: 'instrument', type: 'wavetable', version: '1.0.0', description: 'Sharp wavetable pluck', author: 'Max Booster', grade: 'A', oscillators: [{ type: 'sawtooth', detune: 0, gain: 0.8 }], envelope: { attack: 0.001, decay: 0.2, sustain: 0, release: 0.15 }, parameters: [{ id: 'position', name: 'Wave Position', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { position: 0.5, attack: 0.8, volume: 0.8 } };

export default MbWtPluckPlugin;
